src/lib/rollback_cmdgen.c

```diff
@@ -0,0 +1,252 @@
+                        "make sure root container has ipr2cgen:cmd-add\n",
+                __func__);
+        return NULL;
+    }
+    if (get_extension(CMD_DELETE_EXT, root_node, &oper2cmd_prefix[DELETE_OPR]) != EXIT_SUCCESS) {
+        fprintf(stderr, "%s: cmd-delete extension is missing from root container "
+                        "make sure root container has ipr2cgen:cmd-delete\n",
+                __func__);
+        return NULL;
+    }
+    if (get_extension(CMD_UPDATE_EXT, root_node, &oper2cmd_prefix[UPDATE_OPR]) != EXIT_SUCCESS) {
+        fprintf(stderr, "%s: ipr2cgen:cmd-update extension is missing from root container "
+                        "make sure root container has ipr2cgen:cmd-update\n",
+                __func__);
+        return NULL;
+    }
+
+    return oper2cmd_prefix; // Returning the updated array on success
+}
+
+int lyd_leaf2arg(struct lyd_node *element, bool only_lyd_list_keys, char **arg_name, char **arg_value) {
+    if (only_lyd_list_keys && !lysc_is_key(element->schema)){
+        *arg_name = NULL;
+        *arg_value = NULL;
+        return EXIT_SUCCESS;
+    }
+
+    /* set arg_name */
+    if (get_extension(VALUE_ONLY_EXT, element, NULL) == EXIT_SUCCESS) {
+            *arg_name = NULL;
+    } else {
+        *arg_name = strdup(element->schema->name);
+    }
+    
+    /* set arg_value */
+    if (element->schema->nodetype == LYS_LEAF){
+        LY_DATA_TYPE type = ((struct lysc_node_leaf *)element->schema)->type->basetype;
+        if (type == LY_TYPE_IDENT) {
+            *arg_value = strdup(get_identityref_value(lyd_get_value(element)));
+        } else {
+            *arg_value = strdup(lyd_get_value(element));
+        }
+    }
+    
+    //printf("a-arg_name %s\na-arg_value %s\n", *arg_name, *arg_value);
+    
+    return EXIT_SUCCESS;
+}
+
+int lyd_leaf2rollback_arg(struct lyd_node *element, bool only_lyd_list_keys, char **arg_name, char **arg_value) {
+    
+    if (only_lyd_list_keys && !lysc_is_key(element->schema)){
+        *arg_name = NULL;
+        *arg_value = NULL;
+        return EXIT_SUCCESS;
+    }
+
+    /* set arg_name */
+    if (get_extension(VALUE_ONLY_EXT, element, NULL) == EXIT_SUCCESS) {
+            *arg_name = NULL;
+    } else {
+        *arg_name = strdup(element->schema->name);
+    }
+    
+    /* set arg_value */
+
+    if (element->schema->nodetype == LYS_LEAF || element->schema->nodetype == LYS_LEAFLIST){
+        
+        struct lyd_meta *meta_prev_value, *meta_default_value;
+        /* "orig-value" metadata contains the previous value */
+        meta_prev_value = lyd_find_meta(element->meta, NULL, "yang:orig-value");
+        /* "orig-default" holds the previous default flag value */
+        meta_default_value = lyd_find_meta(element->meta, NULL, "yang:orig-default");
+
+
+
+        if (meta_prev_value)
+            printf("node name %s, meta %s\n",element->schema->name,lyd_get_meta_value(meta_prev_value));
```

Comment:
why do you need such printf() ? debugging ?

Reply:
yes, just debugging, many things in this file will change, there is a lot of code duplication with cmdgen.c
the purpose of this WIP PR is to show the code structure so that we agree on one way of structing the code, then we can work on removing duplications